mithril-client/src/certificate_client/api.rs

```diff
@@ -0,0 +1,83 @@
+use async_trait::async_trait;
+use mithril_common::logging::LoggerExtensions;
+use std::sync::Arc;
+
+use crate::aggregator_client::AggregatorClient;
+use crate::certificate_client::fetch::InternalCertificateRetriever;
+use crate::certificate_client::{fetch, verify};
+use crate::{MithrilCertificate, MithrilCertificateListItem, MithrilResult};
+
+/// Aggregator client for the Certificate
+pub struct CertificateClient {
+    pub(super) aggregator_client: Arc<dyn AggregatorClient>,
+    pub(super) retriever: Arc<InternalCertificateRetriever>,
+    pub(super) verifier: Arc<dyn CertificateVerifier>,
+}
+
+impl CertificateClient {
+    /// Constructs a new `CertificateClient`.
+    pub fn new(
+        aggregator_client: Arc<dyn AggregatorClient>,
+        verifier: Arc<dyn CertificateVerifier>,
+        logger: slog::Logger,
+    ) -> Self {
+        let logger = logger.new_with_component_name::<Self>();
+        let retriever = Arc::new(InternalCertificateRetriever::new(
+            aggregator_client.clone(),
+            logger,
+        ));
+
+        Self {
+            aggregator_client,
+            retriever,
+            verifier,
+        }
+    }
+
+    /// Fetch a list of certificates
+    pub async fn list(&self) -> MithrilResult<Vec<MithrilCertificateListItem>> {
+        fetch::list(self).await
+    }
+
+    /// Get a single certificate full information from the aggregator.
+    pub async fn get(&self, certificate_hash: &str) -> MithrilResult<Option<MithrilCertificate>> {
+        fetch::get(self, certificate_hash).await
+    }
+
+    /// Validate the chain starting with the certificate with given `certificate_hash`, return the certificate if
+    /// the chain is valid.
+    ///
+    /// This method will fail if no certificate exists for the given `certificate_hash`.
+    pub async fn verify_chain(&self, certificate_hash: &str) -> MithrilResult<MithrilCertificate> {
+        verify::verify_chain(self, certificate_hash).await
+    }
+}
+
+/// API that defines how to validate certificates.
+#[cfg_attr(test, mockall::automock)]
+#[cfg_attr(target_family = "wasm", async_trait(?Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait)]
+pub trait CertificateVerifier: Sync + Send {
+    /// Validate the chain starting with the given certificate.
+    async fn verify_chain(&self, certificate: &MithrilCertificate) -> MithrilResult<()>;
+}
+
+#[cfg(feature = "unstable")]
+/// API that defines how to cache certificates validation results.
+#[cfg_attr(test, mockall::automock)]
+#[cfg_attr(target_family = "wasm", async_trait(?Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait)]
+pub trait CertificateVerifierCache: Sync + Send {
+    /// Store a validated certificate hash and its parent hash in the cache.
+    async fn store_validated_certificate(
```

Comment:
I will change the API to take a `entities::Certificate` so the name will be relevant (this will also fit with a refactor where we would move this system to common).

Reply:
After trying this change It's more difficult that I expected. The `mithril-client` library don't expose directly or with an alias the `entities::Certificate`, so using it here would mean either:

- third party implementers must add a dependency to `mithril-common`, as we would need to do in `mithril-client-wasm`
- we expose the entity in the client lib, adding a confusion with the already exposed `CertificateMessage` (exposed with the `MithrilCertificate` alias)

Another option would be to add a new type like `CertificateToCache` which would contains the properties that a cache system implementers could use.

The best option in my opinion is to keep it as is (with an potential rename) or to add a new type.